xds/src/main/java/io/grpc/xds/ServerXdsClient.java

```diff
@@ -0,0 +1,240 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Supplier;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.envoyproxy.envoy.config.core.v3.Address;
+import io.envoyproxy.envoy.config.listener.v3.FilterChain;
+import io.envoyproxy.envoy.config.listener.v3.FilterChainMatch;
+import io.envoyproxy.envoy.config.listener.v3.Listener;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.xds.EnvoyProtoData.Node;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * XdsClient implementation for server side usages.
+ */
+final class ServerXdsClient extends AbstractXdsClient {
+
+  // Longest time to wait, since the subscription to some resource, for concluding its absence.
+  @VisibleForTesting
+  static final int INITIAL_RESOURCE_FETCH_TIMEOUT_SEC = 15;
+  private final SynchronizationContext syncContext;
+  @Nullable
+  private ListenerWatcher listenerWatcher;
+  private int listenerPort = -1;
+  @Nullable
+  private ScheduledHandle ldsRespTimer;
+
+  ServerXdsClient(
+      XdsChannel channel,
+      Node node,
+      SynchronizationContext syncContext,
+      ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Supplier<Stopwatch> stopwatchSupplier) {
+    super(channel, node, timeService, backoffPolicyProvider, stopwatchSupplier.get());
+    this.syncContext = checkNotNull(syncContext, "syncContext");
+  }
+
+  @Override
+  void watchListenerData(int port, ListenerWatcher watcher) {
+    checkState(listenerWatcher == null, "ListenerWatcher already registered");
+    listenerWatcher = checkNotNull(watcher, "watcher");
+    checkArgument(port > 0, "port needs to be > 0");
+    this.listenerPort = port;
+    getLogger().log(XdsLogLevel.INFO, "Started watching listener for port {0}", port);
+    updateNodeMetadataForListenerRequest(port);
+    adjustResourceSubscription(ResourceType.LDS);
+    if (isInBackoff()) {
```

Comment:
Should it be `if (!isInBackoff())` ?

Reply:
Oops. Sorry, my bad.

Reply:
I highly recommend having a  unit test that checks that `ldsRespTimer` is not created/set if the client is in backoffMode to catch such things.

Reply:
That needs to be done for client and server side respectively. On the client side we have [`streamClosedAndRetryRaceWithAddRemoveWatchers`](https://github.com/grpc/grpc-java/blob/0e7cd05bf47d9a408bb834c10558789d231e3175/xds/src/test/java/io/grpc/xds/XdsClientImplTest2.java#L1433). But it doesn't verify the resource subscribe timers. It can be added easily into that test case. But maybe not now. This is an existing behavior, so we should not expect anything related to it to be changed even though it is missing some tests. Also, can you take care of the server side if you want make server side test more robust?